### PR TITLE
Verify weight

### PR DESCRIPTION
**Pull request by AWice**

### CODE REVIEW

**Review by mariano54 — CHANGES_REQUESTED** [review]
This code prefers Blockchain blocks over the new chain's blocks, which means it will not get the correct difficulty in a longer reorg. 

Also it's not checking proofs of time.

At the very least it needs to calculate challenges, ips, and check proof of time for each block. Perhaps we should also check prev_header_hash in the header. 

Finally, we should add tests to make sure it's doing these things correctly.

**Comment by mariano54** [comment]
Can you make the PR into 1.4 instead of 1.3? Hard to see the changes here.

**Review by mariano54 — CHANGES_REQUESTED** [review]
Can you also write a test or two for this function? You can model it after the reorg tests in blockchain.py. And try running it with the testnet chain to see how long it takes? I was just thinking that if someone closes the node while this is running, it won't stop (which is problematic if it takes more than 15 minutes). We could check in each execution of the loop whether the node has closed. I can help you with this if you want, you have to pass in a callback from full node into the verify_weight function. 

Another potential optimization is to only do like 100 headers at a time, to reduce memory requirements, but we can probably do that later.

File: src/blockchain.py [hunk 1]
```diff
@@ -187,12 +187,51 @@ def find_fork_point(self, alternate_chain: List[bytes32]) -> uint32:
...
         Used to calculate the number of iterations.
+
+        Including a beanstalk pretends that the given chain of header blocks
+        (in increasing order of height) overwrites the current blockchain for
+        the purpose of computing the next difficulty.  You must include the
+        height of the queried block to include a beanstalk.
         """
-        block: HeaderBlock = self.header_blocks[header_hash]
+
+        def get_header_block_from_height_or_hash(
```
[the remark this answers is not in the document]
Comment (AWice) [reply src/blockchain.py]: I thought to combine them initially, but sometimes you only want difficulty in the recursive calls.  So I think they would be split.

**Comment by mariano54** [comment]
Please change base branch to 1.5.

Also we should use the pre_validation code that straya added to avoid having to validate all headers twice. I'm kind of hesitant of merging this at this point, since it will make syncing really slow, and we will be merging integration branch relatively soon (which has many changes to the format which impact this PR).

**Comment by mariano54** [comment]
Working on including part of this PR into to ms.sync branch

**Comment by mariano54** [comment]
This is replaced by: https://github.com/Chia-Network/chia-blockchain/pull/218

**Pull request closed without merging**